x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java

```diff
@@ -311,38 +304,50 @@ public void expandJobIds(String expression, boolean allowNoJobs, ActionListener<
     }
 
     /**
-     * Mark the job as being deleted. First looks in the cluster state for the
-     * job configuration then the index
+     * Mark the job as being deleted. First looks in the ml-config index
+     * for the job configuration then the clusterstate
      *
      * @param jobId     To to mark
      * @param force     Allows an open job to be marked
      * @param listener  listener
      */
     public void markJobAsDeleting(String jobId, boolean force, ActionListener<Boolean> listener) {
-        if (ClusterStateJobUpdate.jobIsInClusterState(clusterService.state(), jobId)) {
-            ClusterStateJobUpdate.markJobAsDeleting(jobId, force, clusterService, listener);
-        } else {
-            jobConfigProvider.markJobAsDeleting(jobId, listener);
-        }
+        jobConfigProvider.markJobAsDeleting(jobId, ActionListener.wrap(
+                listener::onResponse,
+                e -> {
+                    if (e.getClass() == ResourceNotFoundException.class) {
+                        // is the config in the cluster state
+                        if (ClusterStateJobUpdate.jobIsInClusterState(clusterService.state(), jobId)) {
+                            ClusterStateJobUpdate.markJobAsDeleting(jobId, force, clusterService, listener);
+                            return;
+                        }
+                    }
+                    listener.onFailure(e);
+                }
+        ));
+
     }
 
     /**
-     * First try to delete the job from the cluster state, if it does not exist
-     * there try to  delete the index job.
+     * First try to delete the job document from ml-config, if it does not exist
+     * there try to the clusterstate.
      *
      * @param request   The delete job request
      * @param listener  Delete listener
      */
     public void deleteJob(DeleteJobAction.Request request, ActionListener<Boolean> listener) {
-
-        if (ClusterStateJobUpdate.jobIsInClusterState(clusterService.state(), request.getJobId())) {
-            ClusterStateJobUpdate.deleteJob(request, clusterService, listener);
-        } else {
-            jobConfigProvider.deleteJob(request.getJobId(), false, ActionListener.wrap(
-                    deleteResponse -> listener.onResponse(Boolean.TRUE),
-                    listener::onFailure
-            ));
-        }
+        jobConfigProvider.deleteJob(request.getJobId(), false, ActionListener.wrap(
+                deleteResponse -> {
+                    if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                        if (ClusterStateJobUpdate.jobIsInClusterState(clusterService.state(), request.getJobId())) {
+                            ClusterStateJobUpdate.deleteJob(request, clusterService, listener);
+                        }
+                    } else {
+                        listener.onResponse(deleteResponse.getResult() == DocWriteResponse.Result.DELETED);
```

Comment:
This is interesting, I understand the change as now we are deleting a document and not just something from the cluster state. However, looking at how our other deletions work that are deleting indexes, they just always return true. 

Are we agreeing now that we should modify the `AcknowledgedResponse` based on the success of the deletion?

Right now, the `deleteDatafeed` method asserts that it is deleted, I guess that throws an exception if the doc is not deleted, where as this simply changes the `true` to a `false` in the `AcknowledgedResponse`. I think these methods should be consistent. 

https://github.com/elastic/elasticsearch/blob/a7c78ee541124c3e839ef18c58aa79b5ef1ba041/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/persistence/DatafeedConfigProvider.java#L232-L233

Reply:
`JobConfigProvider#deleteJob()` takes an `errorIfMissing` parameter but `DatafeedConfigProvider#deleteDatafeedConfig()` does not. There are inconsistencies here and the value returned by the listener isn't actually checked. I'll try to clean it up in this PR but if it becomes a big change it will be better to raise a separate PR - one I can easily port to the master feature branch as this PR will not be merged into master (by v7 all jobs will run from index configs).

Assertions are enabled during testing `java -ea ...` but most people will not enable them in production, here they serve as a sanity check because the document should either be not found or deleted anything else is unexpected.

Reply:
Ah, roger.

Reply:
Looking at the code there are refactorings to be made, some functions take a `errorIfMissing` parameter but are only called with it set to false, also there are some functions that are not called anymore (`JobConfigProvider.getJobs`). I'll make those changes in a separate PR